pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -42,48 +41,70 @@
 
 /**
  * This basic {@code BroadcastJoinOperator} implement a basic broadcast join algorithm.
+ * This algorithm assumes that the broadcast table has to fit in memory since we are not supporting any spilling.
  *
+ * For left join, inner join, right join and full join,
  * <p>It takes the right table as the broadcast side and materialize a hash table. Then for each of the left table row,
  * it looks up for the corresponding row(s) from the hash table and create a joint row.
  *
  * <p>For each of the data block received from the left table, it will generate a joint data block.
- *
- * We currently support left join, inner join and semi join.
+ * We currently support left join, inner join, right join and full join.
  * The output is in the format of [left_row, right_row]
  */
 // TODO: Move inequi out of hashjoin. (https://github.com/apache/pinot/issues/9728)
 public class HashJoinOperator extends BaseOperator<TransferableBlock> {
+  private class BroadcastRows {
+    public BroadcastRows(List<Object[]> rows, boolean hasMatch) {
+      _rows = rows;
+      _hasMatch = hasMatch;
+    }
```

Comment:
i am not sure what's the goal here. but higher level question is do we plan to support inequality outer join conditions. 
1. if we do not (e.g. only equality outer join): then there's no need to encapsulate these, simply create another hashmap might be better. 
    - if the join type is LEFT, there' no need to maintain this boxed object and the other HashMap<Key, Boolean> will be empty
    - if the join type is RIGHT/FULL the sole addition is on the HashMap
2. if we want to support inequality join condition, then the hasMap should be attached to every row otherwise the results wont be correct.

Reply:
alternatively we can also maintain 2 hashmaps
```
private final HashMap<Key, List<Object[]>> _unmatchedHashTable;
private final HashMap<Key, List<Object[]>> _matchedHashTable;
```
and for the hash join, simply loop through both 
1. for LEFT and INNER join, remain the same logic (the _matchHashTable will always be empty and thus no extra cost)
2. for RIGHT and FULL join, move a row from _unmatch to _match onc the equality and inequality conditions are applied successfully. 
    - at the very end when a EOS block is reached on the left. simply output the right _umatched rows

Reply:
This doesn't work since left join has a bug when there is inequi condition. we have to record _unmatchedLeftRows for left join and full join.

For right join and full join, we have to remember matched rows to output unmatched ones.
I used  HashMap<Key, HashSet<Integer>> _matchedRightRows because we have to use this to look up unmatched rows.

PTAL at the new implementation.